### PR TITLE
Move static KEM public key to new PublicBox struct

### PR DESCRIPTION
**Pull request by pqcfox**
This PR moves the McEliece public key from secret heap memory to public heap memory using a new `PublicBox` struct, lowering the amount of memfd memory used per peer when using memfd-based allocation.

(Original discussion for this change took place in #341.)

### CODE REVIEW

File: rosenpass/src/protocol.rs [hunk 1]
```diff
@@ -548,7 +548,7 @@ impl CryptoServer {
     pub fn pidm(&self) -> Result<PeerId> {
         Ok(Public::new(
             hash_domains::peerid()?
-                .mix(self.spkm.secret())?
+                .mix(&*self.spkm)?
```
Comment (koraa) [inline rosenpass/src/protocol.rs:551]: Looks all good…how about using .deref() or .borrow() here (as appropriate) to avoid this magic looking incantation?

**Review by koraa — APPROVED** [review]

File: secret-memory/src/public.rs [hunk 2]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+    }
+
+    /// Create a random initialized [PublicBox]
+    pub fn random() -> Self {
+        mutating(Self::zero(), |r| r.randomize())
+    }
+
+    /// Randomize all bytes in an existing [PublicBox]
+    pub fn randomize(&mut self) {
+        (**self).try_fill(&mut crate::rand::rng()).unwrap()
```
Comment (koraa) [inline secret-memory/src/public.rs:314]: This also looks pretty magic…is there a method alternative for this?

File: secret-memory/src/public.rs [hunk 3]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+    /// Create a zero initialized [PublicBox]
+    pub fn zero() -> Self {
+        Self {
+            inner: Box::new(Public::zero()),
+        }
+    }
+
+    /// Create a random initialized [PublicBox]
+    pub fn random() -> Self {
+        mutating(Self::zero(), |r| r.randomize())
```
Comment (koraa) [inline secret-memory/src/public.rs:309]: I think this can be based on Public::random

File: secret-memory/src/public.rs (hunk 2 above)
Comment (koraa) [inline secret-memory/src/public.rs:314]: Also, this can be based on Public::randomize

File: secret-memory/src/public.rs [hunk 4]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+
+impl<const N: usize> DerefMut for PublicBox<N> {
+    fn deref_mut(&mut self) -> &mut [u8; N] {
+        &mut *self.inner
+    }
+}
+
+impl<const N: usize> Borrow<[u8]> for PublicBox<N> {
+    fn borrow(&self) -> &[u8] {
+        &**self
```
Comment (koraa) [inline secret-memory/src/public.rs:346]: These would also look a lot less magic with method style incantations

File: secret-memory/src/public.rs [hunk 5]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+    fn borrow_mut(&mut self) -> &mut [u8] {
+        &mut **self
+    }
+}
+
+impl<const N: usize> LoadValue for PublicBox<N> {
+    type Error = anyhow::Error;
+
+    fn load<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let mut v = Self::random();
```
Comment (koraa) [inline secret-memory/src/public.rs:360]: This can be based on the Public impl for LoadValue

File: secret-memory/src/public.rs [hunk 6]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+        fopen_r(path)?.read_exact_to_end(&mut *v)?;
+        Ok(v)
+    }
+}
+
+impl<const N: usize> StoreValue for PublicBox<N> {
+    type Error = anyhow::Error;
+
+    fn store<P: AsRef<Path>>(&self, path: P) -> anyhow::Result<()> {
+        std::fs::write(path, **self)?;
```
Comment (koraa) [inline secret-memory/src/public.rs:370]: This can be based on the Public impl of StoreValue

File: secret-memory/src/public.rs [hunk 7]
```diff
@@ -277,3 +277,152 @@ mod tests {
...
+}
+
+impl<const N: usize> LoadValueB64 for PublicBox<N> {
+    type Error = anyhow::Error;
+
+    fn load_b64<const F: usize, P: AsRef<Path>>(path: P) -> Result<Self, Self::Error>
+    where
+        Self: Sized,
+    {
+        let mut f = [0u8; F];
```
Comment (koraa) [inline secret-memory/src/public.rs:382]: This can be based on the Public impl of LoadValueB64

**Review by koraa — CHANGES_REQUESTED** [review]
Some of the code seems unnecessarily complex and copy/paste based. Lets write as little code as possible here.

**Comment by koraa** [comment]
@pqcfox I simplified a few of the dereferencing incantations in PublicBox

File: secret-memory/src/public.rs (hunk 7 above)
Comment (koraa) [reply secret-memory/src/public.rs:382]: OK so there is actually some good reason to have this be a separate implementation: Avoiding large stack allocation.

Unfortunately the `let mut f = [0u8; F];` will break this. If we replace this with a heap allocation this all will make sense again!

**Review by koraa — CHANGES_REQUESTED** [review]

**Comment by koraa** [comment]
> ## [Codecov](https://app.codecov.io/gh/rosenpass/rosenpass/pull/350?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass) Report
> 
> Attention: Patch coverage is `45.09804%` with `56 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 73.92%. Comparing base [(`7c54a37`)](https://app.codecov.io/gh/rosenpass/rosenpass/commit/7c54a37618dbdfe639f19401809d4bd594d1d64e?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass) to head [(`2eefab2`)](https://app.codecov.io/gh/rosenpass/rosenpass/commit/2eefab265bf65b7c8904849c6a5b3d108e1ec61d?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass).
> 
> [Files](https://app.codecov.io/gh/rosenpass/rosenpass/pull/350?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass) 	Patch % 	Lines
> [secret-memory/src/public.rs](https://app.codecov.io/gh/rosenpass/rosenpass/pull/350?src=pr&el=tree&filepath=secret-memory%2Fsrc%2Fpublic.rs&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass#diff-c2VjcmV0LW1lbW9yeS9zcmMvcHVibGljLnJz) 	31.70% 	[56 Missing ⚠️ ](https://app.codecov.io/gh/rosenpass/rosenpass/pull/350?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass)
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/rosenpass/rosenpass/pull/350?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=rosenpass).

@pqcfox Could you also make sure to reuse the `Public` tests?

**Comment by pqcfox** [comment]
@koraa thanks so much! I'll continue simplifying and add the coverage today.

**Comment by pqcfox** [comment]
Perfect, made the changes requested @koraa! A couple of small notes:

- For one, the stack allocation comment you made was spot on--in particular the stack overflows during integration tests if I define `PublicBox::load` in terms of `Public::load`, so I wrote it in terms of `PublicBox::zero` instead.
- For `PublicBox::LoadValueB64` and your comment on that it might be useful to avoid excessive stack usage, I opted to use a `vec!` to move the backing memory for the scratch variable `t` to the heap. I also considered `Box::new([0u8; N])`, but I realized that that would still end up allocating stack space, and I wanted to avoid having to use unstable features like `Box::new_zeroed()` if possible.

Let me know your thoughts, thanks so much for the excellent feedback!

**Comment by koraa** [comment]
@pqcfox Its great :)

**Comment by pqcfox** [comment]
Thanks! Excited to have this merged!

**Pull request merged**